### ISSUE TITLE
bootstrap submodule sync overwrites local changes

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#103485]
When bootstrap updates a submodule, it performs a hard reset, discarding all local changes: https://github.com/rust-lang/rust/blob/1481fd964bac3c750c7e1b21206fdaa60346c456/src/bootstrap/lib.rs#L631-L632

This was a nasty surprise when I was editing my local `src/llvm-project`, and then did a `git pull` and `./x.py build`.
Fortunately, I only had some extra debugging prints added, but there was no indication that my local changes would be overwritten.

We should require explicit confirmation before potentially deleting local changes like this.

**Comment by chenyukang** [comment]
There is an option in `config.toml`:
Automatic submodule handling can still be disabled with `build.submodules = false`.

We'd better get confirmation if there is no explicit config parameter get from `config.toml`?

**Comment by nagisa** [comment]
This has been a frequent source of surprises for me as well going back all the way to the addition of `x.py`. More generally I feel it is important that we figure out what the out-of-the-box experience is tailored for – the developers or people building rustc for distribution. Today this is a mixed bag.

**Comment by jyn514** [comment]
I think a simple fix would be to use git stash to save changes before doing a hard reset; if that causes issues for some reason, we can remove all the `--force` flags so that we don't overwrite local changes and give a hard error instead.

**Comment by pratushrai0309** [comment]
Can I work on this issue? Though I need Some guidance I am new to the project

**Comment by jyn514** [comment]
@pratushrai0309 first, reproduce the original issue so you can be sure your changes are having an effect. Then modify the git commands linked in the original post (https://github.com/rust-lang/rust/issues/103485#issue-1421026710) to add `git stash save` before and `git stash apply` after. If that saves your changes then your work is done and you can open a PR; otherwise you'll have to do some experimenting with git to see what works.

**Comment by pratushrai0309** [comment]
> 

So what exactly I should do to reproduce the original Issue. This is my first time contributing to the compiler.

**Comment by pratushrai0309** [comment]
That's what I have done
       `self.run(Command::new("git").args(&["stash", "save"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["stash", "apply"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["reset", "-q", "--hard"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["clean", "-qdfx"]).current_dir(absolute_path));`

**Comment by pratushrai0309** [comment]
That's what I have done
```    self.run(Command::new("git").args(&["stash", "save"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["stash", "apply"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["reset", "-q", "--hard"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["clean", "-qdfx"]).current_dir(absolute_path));```rust

**Comment by pratushrai0309** [comment]
That's what I have done
```rust
       self.run(Command::new("git").args(&["stash", "save"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["stash", "apply"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["reset", "-q", "--hard"]).current_dir(&absolute_path));
        self.run(Command::new("git").args(&["clean", "-qdfx"]).current_dir(absolute_path));
```

**Comment by vincenzopalazzo** [comment]
@pratushrai0309 Maybe a Zulip thread is a better place to show your result and receive more feedback on the solution

**Comment by chenyukang** [comment]
@pratushrai0309 
I think `git apply` will restore the prvious local changes, so if we want to get previous changes, it should added at the ending.

**Comment by pratushrai0309** [comment]
> @pratushrai0309 I think `git stash apply` will restore the prvious local changes, so if we want to get previous changes, it should added at the ending.

Yeah! It was a mistake thanks for pointing out

**Comment by pratushrai0309** [comment]
[https://github.com/rust-lang/rust/pull/103810](url)

**Comment by pratushrai0309** [comment]
#103810

**Comment by pratushrai0309** [comment]
I have done the intended changes and opened a PR #103810

**Comment by pratushrai0309** [comment]
closes #103810

**Comment by pratushrai0309** [comment]
I have done the intended changes and opened a PR #103810 .
Can someone please review it?
